### PR TITLE
check T of mixed&Foo and T of mixed|Foo

### PR DESCRIPTION
**Pull request by schlndh**
Fixes https://phpstan.org/r/a3548a66-d83b-46d4-b418-7dbc038cbeb1 (all var_dumps should be reported).
Fixes https://phpstan.org/r/7ec28c62-5411-412a-a4c8-c7f9253b918d (there should be actual type instead of stdClass).

### CODE REVIEW

File: src/Rules/RuleLevelHelper.php [hunk 1]
```diff
@@ -345,27 +360,33 @@ public function findTypeToCheck(
 		if ($type instanceof MixedType || $type instanceof NeverType) {
 			return new FoundTypeResult(new ErrorType(), [], [], null);
 		}
-		if ($type instanceof StaticType) {
-			$type = $type->getStaticObjectType();
+		if (!$this->newRuleLevelHelper) {
+			if ($isTopLevel && $type instanceof StaticType) {
+				$type = $type->getStaticObjectType();
```
Comment (schlndh) [inline src/Rules/RuleLevelHelper.php:365]: This line hid the following issue (`CallMethodsRuleTest`):
```
Parameter #1 $other of method Test\CollectionWithStaticParam::add() expects static(Test\AppleCollection), Test\AppleCollection given.
```

**Comment by ondrejmirtes** [comment]
1) Looks like this fixed https://github.com/phpstan/phpstan/issues/9009, please add a regression test :)
2) There's a levels test failure https://github.com/phpstan/phpstan-src/actions/runs/8030898804/job/21938683114, you can just re-run that locally and commit the updated JSON file :)

Thank you very much!

File: tests/PHPStan/Rules/Methods/CallStaticMethodsRuleTest.php [hunk 2]
```diff
@@ -250,7 +250,7 @@ public function testCallStaticMethods(): void
 				'Learn more at https://phpstan.org/user-guide/discovering-symbols',
 			],
 			[
-				'Call to an undefined static method static(CallStaticMethods\CallWithStatic)::nonexistent().',
```
Comment (schlndh) [inline tests/PHPStan/Rules/Methods/CallStaticMethodsRuleTest.php:253]: I tried to minimize breaking the previous error messages, and this is the result that I ended up with. IMO `static(...)` is unnecessary information here - the `nonexistent` method has to exist on `CallWithStatic` for the issue to go away.

Should I limit this change to bleeding edge only (e.g. via `newRuleLevelHelper`)?

Comment (ondrejmirtes) [reply tests/PHPStan/Rules/Methods/CallStaticMethodsRuleTest.php:253]: I think it's fine as it is 😊

**Comment by schlndh** [comment]
I thought that I caused a false positive here: https://github.com/drupal/drupal/blob/5c30729a8365dc7e7d649d597a11f170eecff26c/core/modules/migrate/tests/src/Kernel/MigrateTestBase.php#L173
https://github.com/phpstan/phpstan-src/actions/runs/8031005259/job/21938938357?pr=2940

```
  ------ ------------------------------------------------------------------ 
  Line   core/modules/migrate/tests/src/Kernel/MigrateTestBase.php         
 ------ ------------------------------------------------------------------ 
  173    Call to an undefined static method                                
         Drupal\Tests\migrate\Kernel\MigrateTestBase::migrateDumpAlter().  
 ------ ------------------------------------------------------------------ 
```

But it looks like the issue is already present in phpstan, I just changed the error message: https://phpstan.org/r/1475efe6-5aaf-4922-bab9-c4ebacc8d24c

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**